File: go/sort.go
package main

import (
	"bytes"
	"fmt"
	"reflect"
	"strings"
	"time"
	"math/rand"
	"strconv"
)

var rdm *rand.Rand = rand.New(rand.NewSource(time.Now().UnixNano()))

func array(size int, digit int) []int {
	arr := make([]int, size, size)
	for i := range arr {
		arr[i] = rdm.Intn(digit)
	}
	return arr
}
func getSimpleName(i interface{}) string {
	name := reflect.ValueOf(i).Type().String()
	a := strings.Split(name, ".")
	return a[len(a) - 1]
}
func getArrayString(arr []int) string {
	buf := bytes.Buffer{}
	buf.WriteString("[")
	for i := 0; i < len(arr) - 1; i++ {
		buf.WriteString(strconv.Itoa(arr[i]))
		buf.WriteString(", ")
	}
	buf.WriteString(strconv.Itoa(arr[len(arr) - 1]))
	buf.WriteString("]")
	return buf.String()
}
func test(sort Sorter) {
	arr := array(20, 1000)
	sort.sort(arr)
	fmt.Printf("%s\t%s\n", getSimpleName(sort), getArrayString(arr))
}

type Sorter interface {
	sort(arr []int)
}

type BubbleSort struct{}

func (self BubbleSort) sort(arr []int) {
	for i := len(arr) - 1; i >= 0; i-- {
		for j := 0; j < i; j++ {
			if arr[j] > arr[j + 1] {
				tmp := arr[j]
				arr[j] = arr[j + 1]
				arr[j + 1] = tmp
			}
		}
	}
}

type SelectSort struct{}

func (self SelectSort) sort(arr []int) {
	count := len(arr)
	for i := 0; i < count; i++ {
		min := i
		for j := i + 1; j < count; j++ {
			if arr[min] > arr[j] {
				min = j
			}
		}
		if min != i {
			tmp := arr[i]
			arr[i] = arr[min]
			arr[min] = tmp
		}
	}
}

type InsertSort struct{}

func (self InsertSort) sort(arr []int) {
	count := len(arr)
	for i := 1; i < count; i++ {
		j := i - 1
		tmp := arr[i]
		for j >= 0 && arr[j] > tmp {
			arr[j+1] = arr[j]
			j -= 1
		}
		arr[j+1] = tmp
	}
}

type ShellSort struct{}

func (self ShellSort) sort(arr []int) {
	count := len(arr)
	gap := count / 2
	for 1 <= gap {
		for i := gap; i < count; i++ {
			j := i - gap
			tmp := arr[i]
			for j >= 0 && arr[j] > tmp {
				arr[j+gap] = arr[j]
				j -= gap
			}
			arr[j+gap] = tmp
		}
		gap /= 2
	}
}

type MergeSort struct{}

func (self MergeSort) sort(arr []int) {
	self.recursive(arr, 0, len(arr)-1)
}

func (self MergeSort) recursive(arr []int, l int, r int)  {
	if l >= r { return }
	m := (r + l) / 2
	self.recursive(arr, l, m)
	self.recursive(arr, m+1, r)
	self.merge(arr, l, m, r)
}

func (self MergeSort) merge(arr []int, l int, m int, r int)  {
	tmp := make([]int, 0, r - l + 1)
	i, j := l, m+1
	for i <= m && j <= r {
		if arr[i] < arr[j] {
			tmp = append(tmp, arr[i])
			i++
		} else {
			tmp = append(tmp, arr[j])
			j++
		}
	}
	for i <= m {
		tmp = append(tmp, arr[i])
		i++
	}
	for j <= r {
		tmp = append(tmp, arr[j])
		j++
	}
	for i, v := range tmp {
		arr[l+i] = v;
	}
}

type QuickSort struct{}

func (self QuickSort) sort(arr []int) {
	self.recursive(arr, 0, len(arr)-1)
}

func (self QuickSort) recursive(arr []int, l int, r int)  {
	if l >= r { return }
	m := self.partition(arr, l, r)
	self.recursive(arr, l, m-1)
	self.recursive(arr, m+1, r)
}

func (self QuickSort) partition(arr []int, l int, r int) int {
	m := l
	for i := l; i < r; i++ {
		if arr[i] < arr[r] {
			temp := arr[i]
			arr[i] = arr[m]
			arr[m] = temp
			m++
		}
	}
	temp := arr[r]
	arr[r] = arr[m]
	arr[m] = temp
	return m
}

func main() {
	test(BubbleSort{})
	test(SelectSort{})
	test(InsertSort{})
	test(ShellSort{})
	test(MergeSort{})
	test(QuickSort{})
}

File: php/sort.php
<?php
//https://tool.lu/coderunner/
// 获得一个随机数组
function random_array($size=10, $digit=10) {
    $arr = range(0, $size-1);
    for ($i=0; $i < $size; $i++) {
        $arr[$i] = rand(0, $digit-1);
    }
    return $arr;
}

function test($sorter) {
    $arr = random_array(20, 1000);
    $sorter->sort($arr);
    echo get_class($sorter).'&nbsp;&nbsp;';
    foreach ($arr as $k => $v) {
        echo ($k == 0 ? '[' : '').$v.($k == count($arr)-1 ? ']<br>' : ', ');
    }
}

interface Sorter {
    public function sort(&$arr);
}
// 冒泡排序
class BubbleSort implements Sorter {
    public function sort(&$arr) {
        for ($i=count($arr)-1; $i >= 0; $i--) {
            for ($j=0; $j < $i; $j++) {
                if ($arr[$j] > $arr[$j+1]) {
                    $tmp = $arr[$j];
                    $arr[$j] = $arr[$j+1];
                    $arr[$j+1] = $tmp;
                }
            }
        }
    }
}
// 选择排序
class SelectSort implements Sorter {
    public function sort(&$arr) {
        for ($i = 0; $i < count($arr); $i++) {
            $min = $i;
            for ($j = $i + 1; $j < count($arr); $j++) {
                if ($arr[$j] < $arr[$min])
                    $min = $j;
            }
            if ($min != $i) {
                $temp = $arr[$i];
                $arr[$i] = $arr[$min];
                $arr[$min] = $temp;
            }
        }
    }
}
// 插入排序
class InsertSort implements Sorter {
    public function sort(&$arr) {
        for ($i = 1; $i < count($arr); $i++) {
            $j = $i - 1;
            $tmp = $arr[$i];
            while ($j >= 0 && $tmp < $arr[$j]) {
                $arr[$j + 1] = $arr[$j];
                $j--;
            }
            $arr[$j + 1] = $tmp;
        }
    }
}
// 希尔排序
class ShellSort implements Sorter {
    public function sort(&$arr) {
        $n = count($arr);
        $gap = (int)($n / 2);
        while (1 <= $gap) {
            for ($i = $gap; $i < $n; $i++) {
                $j = $i - $gap;
                $tmp = $arr[$i];
                while ($j >= 0 && $tmp < $arr[$j]) {
                    $arr[$j + $gap] = $arr[$j];
                    $j -= $gap;
                }
                $arr[$j + $gap] = $tmp;
            }
            $gap = (int)($gap / 2);
        }
    }
}
// 归并排序
class MergeSort implements Sorter {
    public function sort(&$arr) {
        $this->recursive($arr, 0, count($arr) - 1);
    }
    private function recursive(&$arr, $l, $r) {
        if ($l >= $r) return;
        $m = (int)(($r + $l) / 2);
        $this->recursive($arr, $l, $m);
        $this->recursive($arr, $m+1, $r);
        $this->merge($arr, $l, $m, $r);
    }
    private function merge(&$arr, $l, $m, $r) {
        $i = $l; $j = $m + 1; $tmp = array();
        while ($i <= $m && $j <= $r) array_push($tmp, ($arr[$i] < $arr[$j]) ? $arr[$i++] : $arr[$j++]);
        while ($i <= $m) array_push($tmp, $arr[$i++]);
        while ($j <= $r) array_push($tmp, $arr[$j++]);
        foreach ($tmp as $i => $v) $arr[$l + $i] = $v;
    }
}
// 快速排序
class QuickSort implements Sorter {
    public function sort(&$arr) {
        $this->recursive($arr, 0, count($arr) - 1);
    }
    private function recursive(&$arr, $l, $r) {
        if ($l >= $r) return;
        $m = $this->partition($arr, $l, $r);
        $this->recursive($arr, $l, $m-1);
        $this->recursive($arr, $m+1, $r);
    }
    private function partition(&$arr, $l, $r) {
        $m = $l;
        for ($u = $l; $u < $r; $u++) {
            if ($arr[$u] <= $arr[$r]) {
                $temp = $arr[$u];
                $arr[$u] = $arr[$m];
                $arr[$m] = $temp;
                $m++;
            }
        }
        $temp = $arr[$r];
        $arr[$r] = $arr[$m];
        $arr[$m] = $temp;
        return $m;
    }
}
// 堆排序
class HeapSort implements Sorter {
    public function sort(&$arr) {
        $n = count($arr) - 1;
        for ($i = (int)($n / 2); $i >= 0; $i--) $this->sift($arr, $i, $n);
        for ($i = $n; $i >= 1; $i--) {
            $temp = $arr[$i];
            $arr[$i] = $arr[0];
            $arr[0] = $temp;
            $this->sift($arr, 0, $i - 1);
        }
    }
    private function sift(&$arr, $l, $r) {
        $i = $l; $j = $l * 2 + 1; $tmp = $arr[$l];
        while ($j <= $r) {
            if ($j < $r && $arr[$j] < $arr[$j + 1]) $j++;
            if ($tmp < $arr[$j]) {
                $arr[$i] = $arr[$j];
                $i = $j;
                $j = 2 * $i + 1;
            } else break;
        }
        $arr[$i] = $tmp;
    }
}
// 基数排序
class RadixSort implements Sorter {
    public function sort(&$arr) {
        $n = count($arr); $d = $this->maxbit($arr); $radix = 1;
        $tmp = range(0, $n - 1);
        for ($i = 1; $i <= $d; $i++) {
            $count = array(0, 0, 0, 0, 0, 0, 0, 0, 0, 0);
            for ($j=0; $j < $n; $j++) $count[((int)($arr[$j] / $radix)) % 10]++;
            for ($j=1; $j < 10; $j++) $count[$j] += $count[$j - 1];
            for ($j=$n-1; $j >= 0 ; $j--) $tmp[--$count[((int)($arr[$j] / $radix)) % 10]] = $arr[$j];
            for ($j=0; $j < $n; $j++) $arr[$j] = $tmp[$j];
            $radix *= 10;
        }
    }
    private function maxbit(&$arr) {
        $d = 1; $p = 10;
        foreach ($arr as $i) {
            while ($i >= $p) {
                $p *= 10;
                $d++;
            }
        }
        return $d;
    }
}

test(new BubbleSort());
test(new SelectSort());
test(new InsertSort());
test(new ShellSort());
test(new MergeSort());
test(new QuickSort());
test(new HeapSort());
test(new RadixSort());

?>
